File: apps/desktop/src-tauri/src/commands/embeddings.rs
/*!
 * embeddings.rs — W5-5 Tauri IPC commands for the AI / embeddings cluster.
 *
 * ## Commands
 *   - `embeddings_index_note`  — embed a note and persist it in the store.
 *   - `embeddings_remove_note` — delete the embedding for a renamed/deleted note.
 *   - `embeddings_search`      — semantic nearest-neighbor search.
 *
 * ## Architecture
 *   - `EmbedProvider` (trait) is chosen at runtime from the user's provider
 *     config.  At this layer we always resolve to an `OllamaProvider` unless
 *     an `openai_api_key` is supplied.  Provider selection can be made
 *     configurable via the Settings panel later without touching IPC shape.
 *   - The `EmbeddingStore` lives in `<vault>/.scriptor/embeddings.db`.
 *   - Sealed spans are the caller's responsibility to strip (I-3).  The
 *     front-end must call `vault_strip_sealed` before passing `text`.
 *   - No Scriptor-operated server is used; provider is user-supplied (I-4).
 */

use scriptor_embeddings::{
    EmbedProvider, EmbeddingRecord, EmbeddingStore, NoteEmbedder, OllamaProvider,
};
use serde::Deserialize;
use std::path::PathBuf;
use tauri::command;

// ── Provider config (passed per-call to avoid global shared state) ─────────────

/// Minimal provider configuration the front-end sends with each command.
/// Only one of `ollama_base_url` / `openai_api_key` should be set at a time.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderConfig {
    /// Vault root — used to locate `<vault>/.scriptor/embeddings.db`.
    pub vault_root: String,

    // Ollama (default, local)
    /// e.g. `"http://localhost:11434"` — if set, Ollama is used.
    pub ollama_base_url: Option<String>,
    /// e.g. `"nomic-embed-text"`.
    pub ollama_model: Option<String>,
    /// Expected dimension of the Ollama model (default: 768).
    pub ollama_dimension: Option<usize>,

    // OpenAI
    /// User's API key from the OS keychain.  If set, OpenAI is preferred.
    pub openai_api_key: Option<String>,
    /// e.g. `"text-embedding-3-small"` (default).
    pub openai_model: Option<String>,
}

// ── Helper: open store from vault root ────────────────────────────────────────

fn open_store(vault_root: &str, dimension: usize) -> Result<EmbeddingStore, String> {
    let db_path: PathBuf = [vault_root, ".scriptor", "embeddings.db"].iter().collect();
    EmbeddingStore::open(&db_path, dimension).map_err(|e| e.to_string())
}

// ── Commands ──────────────────────────────────────────────────────────────────

/// Embed `text` (sealed spans pre-stripped by caller) and persist it in the
/// store keyed by `note_path` (vault-relative).
///
/// Uses Ollama if `ollama_base_url` is set, OpenAI otherwise.
#[command]
pub fn embeddings_index_note(
    note_path: String,
    text: String,
    config: ProviderConfig,
) -> Result<(), String> {
    if let Some(api_key) = &config.openai_api_key {
        let model = config
            .openai_model
            .as_deref()
            .unwrap_or("text-embedding-3-small");
        let provider = scriptor_embeddings::OpenAiProvider::new(api_key, model, None);
        let store = open_store(&config.vault_root, provider.dimension())?;
        let embedder = NoteEmbedder { store, provider };
        embedder
            .index_note(&note_path, &text)
            .map_err(|e| e.to_string())
    } else {
        let base_url = config
            .ollama_base_url
            .as_deref()
            .unwrap_or("http://localhost:11434");
        let model = config.ollama_model.as_deref().unwrap_or("nomic-embed-text");
        let dim = config.ollama_dimension.unwrap_or(768);
        let provider = OllamaProvider::new(base_url, model, dim);
        let store = open_store(&config.vault_root, dim)?;
        let embedder = NoteEmbedder { store, provider };
        embedder
            .index_note(&note_path, &text)
            .map_err(|e| e.to_string())
    }
}

/// Remove the embedding record for a deleted or renamed note.
#[command]
pub fn embeddings_remove_note(
    note_path: String,
    vault_root: String,
    dimension: Option<usize>,
) -> Result<(), String> {
    let dim = dimension.unwrap_or(768);
    let store = open_store(&vault_root, dim)?;
    store
        .delete_embedding(&note_path)
        .map_err(|e| e.to_string())
}

/// Semantic nearest-neighbor search.  Returns up to `limit` results sorted by
/// descending cosine similarity.
#[command]
pub fn embeddings_search(
    query: String,
    limit: Option<usize>,
    config: ProviderConfig,
) -> Result<Vec<EmbeddingRecord>, String> {
    let k = limit.unwrap_or(20);

    if let Some(api_key) = &config.openai_api_key {
        let model = config
            .openai_model
            .as_deref()
            .unwrap_or("text-embedding-3-small");
        let provider = scriptor_embeddings::OpenAiProvider::new(api_key, model, None);
        let store = open_store(&config.vault_root, provider.dimension())?;
        let embedder = NoteEmbedder { store, provider };
        embedder.search(&query, k).map_err(|e| e.to_string())
    } else {
        let base_url = config
            .ollama_base_url
            .as_deref()
            .unwrap_or("http://localhost:11434");
        let model = config.ollama_model.as_deref().unwrap_or("nomic-embed-text");
        let dim = config.ollama_dimension.unwrap_or(768);
        let provider = OllamaProvider::new(base_url, model, dim);
        let store = open_store(&config.vault_root, dim)?;
        let embedder = NoteEmbedder { store, provider };
        embedder.search(&query, k).map_err(|e| e.to_string())
    }
}

File: src/bridge/commands/embeddings.ts
/**
 * embeddings.ts — W5-5 TypeScript bridge for the AI / embeddings cluster.
 *
 * All calls go through the Tauri IPC layer.  Sealed spans must be stripped
 * by the caller before passing `text` (I-3).  No Scriptor-operated server
 * is involved; providers are configured by the user (I-4).
 */

import { invoke } from '@tauri-apps/api/core'
import { requireNative } from '../native.ts'

// ── Types ─────────────────────────────────────────────────────────────────────

/**
 * Embedding provider configuration sent with each command.
 * Only one of ollama / openai should be configured at a time.
 */
export interface EmbeddingProviderConfig {
  /** Vault root used to locate <vault>/.scriptor/embeddings.db */
  vaultRoot: string

  // Ollama (local, no key required)
  ollamaBaseUrl?: string    // default: "http://localhost:11434"
  ollamaModel?: string      // default: "nomic-embed-text"
  ollamaDimension?: number  // default: 768

  // OpenAI
  openaiApiKey?: string     // from OS keychain; triggers OpenAI provider
  openaiModel?: string      // default: "text-embedding-3-small"
}

/**
 * A single semantic-search result.
 * `score` is cosine similarity in [0, 1]; higher = more relevant.
 */
export interface EmbeddingRecord {
  notePath: string
  score: number
}

// ── Commands ──────────────────────────────────────────────────────────────────

/**
 * Embed a note and persist it in the embedding store.
 *
 * @param notePath Vault-relative path, e.g. `"Projects/Alpha.md"`.
 * @param text     Plain text with sealed spans stripped (I-3).
 * @param config   Provider + vault config.
 */
export async function embeddingsIndexNote(
  notePath: string,
  text: string,
  config: EmbeddingProviderConfig,
): Promise<void> {
  requireNative()
  return invoke<void>('embeddings_index_note', { notePath, text, config })
}

/**
 * Remove the embedding record for a deleted or renamed note.
 *
 * @param notePath  Vault-relative path of the note to remove.
 * @param vaultRoot Vault root (to locate the embeddings DB).
 * @param dimension Optional; must match the stored dimension. Default 768.
 */
export async function embeddingsRemoveNote(
  notePath: string,
  vaultRoot: string,
  dimension?: number,
): Promise<void> {
  requireNative()
  return invoke<void>('embeddings_remove_note', { notePath, vaultRoot, dimension })
}

/**
 * Semantic nearest-neighbor search over all indexed notes.
 *
 * Returns up to `limit` results sorted by descending cosine similarity.
 * The front-end should merge these with BM25 keyword results using the
 * `notePath` as the join key.
 *
 * @param query  Natural-language query string.
 * @param limit  Max results to return (default 20).
 * @param config Provider + vault config.
 */
export async function embeddingsSearch(
  query: string,
  config: EmbeddingProviderConfig,
  limit?: number,
): Promise<EmbeddingRecord[]> {
  requireNative()
  return invoke<EmbeddingRecord[]>('embeddings_search', { query, limit, config })
}
